File: plugin/src/Logger.cpp
#include "Logger.h"

#include <chrono>
#include <ctime>
#include <iomanip>

Logger::Logger(std::filesystem::path path) {
    out_.open(path, std::ios::app);
}

void Logger::info(const std::string& message) {
    write("INFO", message);
}

void Logger::error(const std::string& message) {
    write("ERROR", message);
}

void Logger::write(const char* level, const std::string& message) {
    std::scoped_lock lock(mutex_);
    if (!out_) {
        return;
    }
    const auto now = std::chrono::system_clock::now();
    const auto time = std::chrono::system_clock::to_time_t(now);
#if defined(_WIN32)
    std::tm tm_buf {};
    localtime_s(&tm_buf, &time);
    std::tm* tm = &tm_buf;
#else
    std::tm* tm = std::localtime(&time);
#endif
    out_ << std::put_time(tm, "%Y-%m-%d %H:%M:%S") << " [" << level << "] " << message << '\n';
    out_.flush();
}

File: plugin/src/logger.rs
use std::fs::{File, OpenOptions};
use std::io::Write;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

pub struct Logger {
    path: PathBuf,
    out: Mutex<Option<File>>,
}

impl Logger {
    pub fn new(path: impl AsRef<Path>) -> Self {
        let path = path.as_ref().to_path_buf();
        let file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&path)
            .ok();
        Self {
            path,
            out: Mutex::new(file),
        }
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    pub fn info(&self, message: impl AsRef<str>) {
        self.write("INFO", message.as_ref());
    }

    pub fn error(&self, message: impl AsRef<str>) {
        self.write("ERROR", message.as_ref());
    }

    fn write(&self, level: &str, message: &str) {
        let Ok(mut guard) = self.out.lock() else {
            return;
        };
        let Some(file) = guard.as_mut() else {
            return;
        };
        let stamp = local_stamp();
        let _ = writeln!(file, "{stamp} [{level}] {message}");
        let _ = file.flush();
    }
}

fn local_stamp() -> String {
    #[cfg(windows)]
    {
        use windows::Win32::System::SystemInformation::GetLocalTime;
        let st = unsafe { GetLocalTime() };
        format!(
            "{:04}-{:02}-{:02} {:02}:{:02}:{:02}",
            st.wYear, st.wMonth, st.wDay, st.wHour, st.wMinute, st.wSecond
        )
    }
    #[cfg(not(windows))]
    {
        let secs = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_secs())
            .unwrap_or(0);
        format!("{secs}")
    }
}
